README.md
# README
## IPK project 2
author: **Dinara Garipova** (xgarip00)  
date: **2023-03-21**
Gitea profile: [xgarip00](https://git.fit.vutbr.cz/xgarip00)

# Task:
## Packet Sniffer
This is a C program for a simple packet sniffer that can capture and display network traffic. The program uses the pcap library to capture packets and parse them. It takes command-line arguments to filter packets based on various criteria such as protocol type, interface, port, etc.



The main function of this program is to parse command-line arguments, set up the filter based on the user's specifications, and start capturing packets. When a packet is captured, it is analyzed to determine its protocol type and other information, which is then displayed on the console.

## Usage
```bash
Usage: ./ipk-sniffer [-i interface | --interface interface] {-p port [--tcp|-t] [--udp|-u]} [--arp] [--icmp4] [--icmp6] [--igmp] [--mld] {-n num}
```
* -i [string]   specify an interface
* -n [integer]  set packet limit (unlimited if not set)
* -p [integer]  set packet port to filter
* -u            filter only UDP packets
* -t            filter only TCP packets

# Requirements
To compile and run this code, you need:

* C compiler (GCC or Clang)
* Linux/Unix system (Ubuntu, Debian, Fedora, macOS)
* Basic knowledge of C programming

## Compilation
This project can be compiled using Makefile:
```bash
make
```
## Compatibility
File can be compiled at 3 different platforms:
* Linux
* Windows **_(not supported)_**
* MacOS


# Implementation
## Content structuring
The program first includes various header files for the libraries and functions it uses. The ProgramArgs structure is defined to store the user's command-line arguments.

The print_interfaces function uses the pcap_findalldevs function from the pcap library to obtain a list of network interfaces on the system and prints them to the console.

The parsing_args function parses the command-line arguments and stores them in the ProgramArgs structure. It checks for various options such as -h or --help for displaying help, -i or --interface for specifying the network interface, -n for setting the packet limit, -p for setting the port number to filter, and various other options for filtering packets based on their protocol type.

The main function first initializes the ProgramArgs structure with default values and calls the parsing_args function to parse the command-line arguments. It then calls the pcap_open_live function from the pcap library to open the specified network interface for capturing packets.

In a loop that runs args.number times, it uses the pcap_next function to get the next packet from the network interface devopen and stores it in the packet variable.
It then extracts the Ethernet header from the packet and prints the source and destination MAC addresses and the frame length.
If the packet is an IPv4 packet, it extracts the IPv4 header from the packet and prints the source and destination IP addresses and the source and destination port numbers (if the packet is TCP or UDP).
If the packet is an ARP packet, it extracts the ARP header from the packet and prints the source and destination IP addresses.
If the packet is an IPv6 packet, it extracts the IPv6 header from the packet and prints the source and destination IP addresses and the source and destination port numbers (if the packet is TCP or UDP).
It then prints a hexadecimal dump of the packet data, 16 bytes per line, with ASCII representation on the right.

## Required libraries
* <string.h> - strcpy(), strcat()
* <stdio.h> - printf()
* <stdlib.h>
* <stdbool.h> - type boolean
* <unistd.h>
* <time.h> - time(), time miliseconds
* <pcap/pcap.h> - access to library functions (etc. pcap_lookupdev())
* <arpa/inet.h> - inet_ntop()

Network Data Structures:
* <netinet/if_ether.h>
* <netinet/ether.h>
* <netinet/ip6.h>
* <netinet/tcp.h>
* <netinet/ip.h>
* <netinet/ip_icmp.h>
* <netinet/udp.h>
* <netinet/in.h>
* <net/ethernet.h>
* <netinet/icmp6.h>

## Functions
### int print_interfaces()
 This function prints the names of all available network interfaces.
 @return 0 if successful, 1 if an error occurs
### int parsing_args(int argc, char *argv[], struct ProgramArgs *args)
Parses command-line arguments and stores them in a struct ProgramArgs.
 @param argc The number of arguments passed to the program.
 @param argv An array of strings representing the arguments passed to the program.
 @param args A pointer to a struct ProgramArgs where the parsed arguments will be stored.
 @return 0 on success, 1 on failure.
### pcap_t* work_with_device(struct ProgramArgs *args)
 @brief capture packets from a network interface, based on user-specified protocols and filters
 @param args pointer to ProgramArgs struct containing user-specified command line arguments
 @return pointer to a pcap_t struct, which can be used to capture packets from the specified network interface
### main(int argc, const char * argv[])
 Description: The main function for the client program Parameters:
 argc - the number of arguments passed to the program
 argv - an array of strings containing the arguments passed to the program Returns: int (0 if successful)

 ## Global Variables
 ### pcap_t* devopen
 A pointer named devopen of type pcap_t. This is likely to be used for opening and reading packets from a network device using the libpcap library.

# Sequence Diagram
   * participant User
   * participant Main
   * participant Parsing_args
   * participant Print_interfaces
   * participant Error
   * participant Work_with_device
   
   ```mermaid
   sequenceDiagram
     User->>Main: Starts the program with input arguments
     activate Main
     Main->>Parsing_args: Checking Program Input Arguments
     Parsing_args->>Error: Throwing an error on invalid arguments
     Parsing_args->>Print_interfaces: If only the interface filter is specified in the arguments
     activate Print_interfaces
     Print_interfaces->>Main: Write out all possible interfaces
     deactivate Print_interfaces

     Main->>Work_with_device: Opening and working with devices

     loop send/recv
         Main->>User: Write out information about each package
     end

     Main->>User: Closes the device
     deactivate Main
   ```

# Testing
## What was tested
All runs tested:
* How the program handles arguments
* How the program composes the string for the filter
* How the program catches the necessary packages
* And how does she write out information about them
## Why it was tested
On some launches, there was a segmentation fault before, so testing was carried out with different filters to make sure the code worked correctly
## How it was tested/what was the testing environment
The program was tested on linux os and nixOS.
For testing, I used a program that sends packages from this repository:
https://github.com/turytsia/vut-ipk-packegen
I entered the arguments I needed, got the output and checked with expectation
## what were the inputs, expected outputs, and actual outputs
#### Input
sudo ./ipk-sniffer -i eth0 --udp
#### expected outputs, and actual outputs
timestamp: 18:39:01.688101
src MAC: 00:15:5D:10:EB:5E
dst MAC: 01:00:5E:7F:FF:FA
frame length: 216 bytes
src IP: 172.25.64.1
dst IP: 239.255.255.250
src port: 17664
dst port: 202

0x0000 01 00 5e 7f ff fa 00 15 5d 10 eb 5e 08 00 45 00  ..^....]..^..E.
0x0010 00 ca 8c 43 00 00 01 11 50 cb ac 19 40 01 ef ff  ...C....P...@...
0x0020 ff fa cf ec 07 6c 00 b6 47 65 4d 2d 53 45 41 52  .....l..GeM-SEAR
0x0030 43 48 20 2a 20 48 54 54 50 2f 31 2e 31 0d 0a 48  CH.*.HTTP/1.1..H
0x0040 4f 53 54 3a 20 32 33 39 2e 32 35 35 2e 32 35 35  OST:.239.255.255
0x0050 2e 32 35 30 3a 31 39 30 30 0d 0a 4d 41 4e 3a 20  .250:1900..MAN:.
0x0060 22 73 73 64 70 3a 64 69 73 63 6f 76 65 72 22 0d  "ssdp:discover".
0x0070 0a 4d 58 3a 20 31 0d 0a 53 54 3a 20 75 72 6e 3a  .MX:.1..ST:.urn:
0x0080 64 69 61 6c 2d 6d 75 6c 74 69 73 63 72 65 65 6e  dial-multiscreen
0x0090 2d 6f 72 67 3a 73 65 72 76 69 63 65 3a 64 69 61  -org:service:dia
0x00a0 6c 3a 31 0d 0a 55 53 45 52 2d 41 47 45 4e 54 3a  l:1..USER-AGENT:
0x00b0 20 47 6f 6f 67 6c 65 20 43 68 72 6f 6d 65 2f 31  .Google.Chrome/1
0x00c0 31 32 2e 30 2e 35 36 31 35 2e 38 36 20 57 69 6e  12.0.5615.86.Win
0x00d0 64 6f 77 73 0d 0a 0d 0a                   dows....

#### Input
 sudo ./ipk-sniffer -i eth0 -n 5
#### expected outputs, and actual outputs
timestamp: 18:17:21.864034
src MAC: 00:15:5D:10:EB:5E
dst MAC: 01:00:5E:7F:FF:FA
frame length: 179 bytes
src IP: 172.25.64.1
dst IP: 239.255.255.250
src port: 17664
dst port: 165
 
0x0000 01 00 5e 7f ff fa 00 15 5d 10 eb 5e 08 00 45 00  ..^....]..^..E.
0x0010 00 a5 8c 15 00 00 04 11 4e 1e ac 19 40 01 ef ff  ........N...@...
0x0020 ff fa ce c3 07 6c 00 91 5a 69 4d 2d 53 45 41 52  .....l..ZiM-SEAR
0x0030 43 48 20 2a 20 48 54 54 50 2f 31 2e 31 0d 0a 48  CH.*.HTTP/1.1..H
0x0040 6f 73 74 3a 20 32 33 39 2e 32 35 35 2e 32 35 35  ost:.239.255.255
0x0050 2e 32 35 30 3a 31 39 30 30 0d 0a 53 54 3a 20 75  .250:1900..ST:.u
0x0060 72 6e 3a 73 63 68 65 6d 61 73 2d 75 70 6e 70 2d  rn:schemas-upnp-
0x0070 6f 72 67 3a 64 65 76 69 63 65 3a 49 6e 74 65 72  org:device:Inter
0x0080 6e 65 74 47 61 74 65 77 61 79 44 65 76 69 63 65  netGatewayDevice
0x0090 3a 31 0d 0a 4d 61 6e 3a 20 22 73 73 64 70 3a 64  :1..Man:."ssdp:d
0x00a0 69 73 63 6f 76 65 72 22 0d 0a 4d 58 3a 20 33 0d  iscover"..MX:.3.
0x00b0 0a 0d 0a                                       ...

timestamp: 18:17:24.875248
src MAC: 00:15:5D:10:EB:5E
dst MAC: 01:00:5E:7F:FF:FA
frame length: 179 bytes
src IP: 172.25.64.1
dst IP: 239.255.255.250
src port: 17664
dst port: 165

0x0000 01 00 5e 7f ff fa 00 15 5d 10 eb 5e 08 00 45 00  ..^....]..^..E.
0x0010 00 a5 8c 16 00 00 04 11 4e 1d ac 19 40 01 ef ff  ........N...@...
0x0020 ff fa ce c3 07 6c 00 91 5a 69 4d 2d 53 45 41 52  .....l..ZiM-SEAR
0x0030 43 48 20 2a 20 48 54 54 50 2f 31 2e 31 0d 0a 48  CH.*.HTTP/1.1..H
0x0040 6f 73 74 3a 20 32 33 39 2e 32 35 35 2e 32 35 35  ost:.239.255.255
0x0050 2e 32 35 30 3a 31 39 30 30 0d 0a 53 54 3a 20 75  .250:1900..ST:.u
0x0060 72 6e 3a 73 63 68 65 6d 61 73 2d 75 70 6e 70 2d  rn:schemas-upnp-
0x0070 6f 72 67 3a 64 65 76 69 63 65 3a 49 6e 74 65 72  org:device:Inter
0x0080 6e 65 74 47 61 74 65 77 61 79 44 65 76 69 63 65  netGatewayDevice
0x0090 3a 31 0d 0a 4d 61 6e 3a 20 22 73 73 64 70 3a 64  :1..Man:."ssdp:d
0x00a0 69 73 63 6f 76 65 72 22 0d 0a 4d 58 3a 20 33 0d  iscover"..MX:.3.
0x00b0 0a 0d 0a                                       ...

timestamp: 18:17:25.646606
src MAC: 00:15:5D:10:EB:5E
dst MAC: 33:33:00:01:00:02
frame length: 134 bytes
src IP: fe80::aea5:e547:d9b6:31c2
dst IP: ff02::1:2
src port: 546
dst port: 547

0x0000 33 33 00 01 00 02 00 15 5d 10 eb 5e 86 dd 60 08  33......]..^..`.
0x0010 cc c1 00 50 11 01 fe 80 00 00 00 00 00 00 ae a5  ...P............
0x0020 e5 47 d9 b6 31 c2 ff 02 00 00 00 00 00 00 00 00  .G..1...........
0x0030 00 00 00 01 00 02 02 22 02 23 00 50 4a 77 01 6a  .......".#.PJw.j
0x0040 40 cc 00 08 00 02 0c 22 00 01 00 0e 00 01 00 01  @......"........
0x0050 28 f7 33 8d 28 cd c4 a0 47 1d 00 03 00 0c 3b 00  (.3.(...G.....;.
0x0060 15 5d 00 00 00 00 00 00 00 00 00 10 00 0e 00 00  .]..............
0x0070 01 37 00 08 4d 53 46 54 20 35 2e 30 00 06 00 06  .7..MSFT.5.0....
0x0080 00 11 00 17 00 18                           ......

timestamp: 18:17:27.885024
src MAC: 00:15:5D:10:EB:5E
dst MAC: 01:00:5E:7F:FF:FA
frame length: 179 bytes
src IP: 172.25.64.1
dst IP: 239.255.255.250
src port: 17664
dst port: 165

0x0000 01 00 5e 7f ff fa 00 15 5d 10 eb 5e 08 00 45 00  ..^....]..^..E.
0x0010 00 a5 8c 17 00 00 04 11 4e 1c ac 19 40 01 ef ff  ........N...@...
0x0020 ff fa ce c3 07 6c 00 91 5a 69 4d 2d 53 45 41 52  .....l..ZiM-SEAR
0x0030 43 48 20 2a 20 48 54 54 50 2f 31 2e 31 0d 0a 48  CH.*.HTTP/1.1..H

Changelog.md
# ChangeLog
## 0.0.1 - 2023-04-14 
>When I started parsing input arguments, I ran into the problem that I needed the filter flags (true or false) to be available from any part of the program.
But I quickly solved this problem by creating a structure that stores all the information about the input arguments
### Added
Initial stage: parsing input arguments
## 0.1.0 - 2023-04-15 
>Then I ran into the problem that when parsing arguments, I compared not only the flags, but also the names of the interfaces and the like. Because of this, errors about an unknown argument were displayed. This problem was also solved very easily, I just skipped one argument where necessary
### Added
Adding flag argument jumping where appropriate
## 0.2.0 - 2023-04-15 
>I also realized that it is necessary to start parsing not from argument 0, but from the first. And rewrite exit() to return, because main is int type program.
### Added
* Start parsing from 1
* Rewrite exit() to return
## 0.5.0 - 2023-04-15
>The hardest bug to fix turned out to be IPv upper-layer protocol and not supported by proto[x] . I still do not understand what the problem is, perhaps in the old version of the compiler. But I had to completely reinstall gcc
### Fixed
Fix ndp and mld flags in filter
## 1.0.0 - 2023-04-17
>The last thing I ran into was a Segmentation fault when no flag other than -i is set. This was fixed by checking for all flags to false and fixing everything to true in this case
### Fixed
Array size fixes to avoid Segmentation fault


ipk-sniffer.c
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <unistd.h>
#include <string.h>
#include <time.h>
#include<pcap/pcap.h>
#include<arpa/inet.h>
#include<netinet/if_ether.h>
#include<netinet/ether.h>
#include<netinet/ip6.h>
#include<netinet/tcp.h>
#include <netinet/ip.h>
#include<netinet/ip_icmp.h>
#include<netinet/udp.h>
#include <netinet/in.h>
#include <net/ethernet.h>
#include <netinet/icmp6.h>

pcap_t* devopen;

struct ProgramArgs
{
    char* interface;
    int port;
    int number;
    bool tcp;
    bool udp;
    bool arp;
    bool icmp4;
    bool icmp6;
    bool igmp;
    bool mld;
    bool ndp;
};

int Error()
{
    exit(1);
    return 1;
}

/**
 * This function prints the names of all available network interfaces.
 * @return 0 if successful, 1 if an error occurs
*/
int print_interfaces()
{
    char error_buffer[PCAP_ERRBUF_SIZE];
        pcap_if_t *interfaces, *device;

        // retrieve the list of network interfaces
    
        if (pcap_findalldevs(&interfaces, error_buffer) == -1) 
        {
            printf("Error in pcap_findalldevs(): %s", error_buffer);
            exit(1);
            return 1;
        }
        
        // print the names of the network interfaces

        for (device = interfaces; device != NULL; device = device->next) {
            printf("%s\n", device->name);
        }

        // free the memory allocated by the list of network interfaces

        pcap_freealldevs(interfaces);
        return 0;
}

/**
 * Parses command-line arguments and stores them in a struct ProgramArgs.
 * @param argc The number of arguments passed to the program.
 * @param argv An array of strings representing the arguments passed to the program.
 * @param args A pointer to a struct ProgramArgs where the parsed arguments will be stored.
 * @return 0 on success, 1 on failure.
*/

int parsing_args(int argc, char *argv[], struct ProgramArgs *args)
{
    // Check if the number of arguments is less than 3 and the first argument is not -h or --help

    if (argc < 3 && strcmp(argv[1], "-h") != 0 && strcmp(argv[1], "--help") != 0)
    {
        // Check if the first argument is not -i or --interface
        if (strcmp(argv[1], "-i") != 0 && strcmp(argv[1], "--interface") != 0)
        {
            printf("Network interfaces not found.\n");
            printf("Usage:\n");
            printf("./ipk-sniffer [-i interface | --interface interface] {-p port [--tcp|-t] [--udp|-u]} [--arp] [--icmp4] [--icmp6] [--igmp] [--mld] {-n num}\n");
            exit(1);
            return 1;
        }
        else{print_interfaces(); exit(0);}
    }
    // Check if the first argument is -h or --help
    else if (strcmp(argv[1], "-h") == 0 || strcmp(argv[1], "--help") == 0)
    {
        printf("Usage:\n");
        printf("./ipk-sniffer [-i interface | --interface interface] {-p port [--tcp|-t] [--udp|-u]} [--arp] [--icmp4] [--icmp6] [--igmp] [--mld] {-n num}\n");
        printf("-i [string]   specify an interface\n");
        printf("-n [integer]  set packet limit (unlimited if not set)\n");
        printf("-p [integer]  set packet port to filter\n");
        printf("-u            filter only UDP packets\n");
        printf("-t            filter only TCP packets\n");
        exit(0);
    }
    // Initializes a variable i to 1 for argument processing.
    int i = 1;
    // Processes all command-line arguments.
    while (i < argc)
    {
        // Checks if the argument is "--tcp" or "-t" and sets the corresponding flag in the struct to true.
        if (strcmp(argv[i], "--tcp") == 0 || strcmp(argv[i], "-t") == 0) 
        {
            // If the flag is already set, it calls the Error() function.
            if (args->tcp){Error();}
            args->tcp = true;
        }
        // Checks if the argument is "--udp" or "-u" and sets the corresponding flag in the struct to true.
        else if (strcmp(argv[i], "--udp") == 0 || strcmp(argv[i], "-u") == 0) 
        {
            // If the flag is already set, it calls the Error() function.
            if (args->udp){Error();}
            args->udp = true;
        }
        // Checks if the argument is "--arp" and sets the corresponding flag in the struct to true.
        else if (strcmp(argv[i], "--arp") == 0) 
        {
            // If the flag is already set, it calls the Error() function.
            if (args->arp){Error();}
            args->arp = true;
        }
        // Checks if the argument is "--icmp4" and sets the corresponding flag in the struct to true.
        else if (strcmp(argv[i], "--icmp4") == 0) 
        {
            if (args->icmp4){Error();}
            args->icmp4 = true;
        }
        // Checks if the argument is "--icmp6" and sets the corresponding flag in the struct to true.
        else if (strcmp(argv[i], "--icmp6") == 0) 
        {
            if (args->icmp6){Error();}
            args->icmp6 = true;
        }
        // Checks if the argument is "--igmp" and sets the corresponding flag in the struct to true.
        else if (strcmp(argv[i], "--igmp") == 0) 
        {
            if (args->igmp){Error();}
            args->igmp = true;
        }
        // Checks if the argument is "--mld" and sets the corresponding flag in the struct to true.
        else if (strcmp(argv[i], "--mld") == 0) 
        {
            if (args->mld){Error();}
            args->mld = true;
        }
        // Checks if the argument is "--ndp" and sets the corresponding flag in the struct to true.
        else if (strcmp(argv[i], "--ndp") == 0) 
        {
            if (args->ndp){Error();}
            args->ndp = true;
        }
        // Checks if the argument is "-i" and sets the interface name in the struct.
        // If the next argument is not a flag, it sets the interface name to the next argument.
        else if (strcmp(argv[i], "-i") == 0) 
        {
            if ((i+1) < argc)
            {
                if (argv[i+1][0] != '-')
                {
                    strcpy(args->interface, argv[i+1]);
                    i++;
                }
                // If the next argument is a flag, it sets the interface name to an empty string.
                else{args->interface = "";}
            }
        }
        // If the -n argument is passed, it checks that it hasn't already been set and that a valid number has been provided.
        else if (strcmp(argv[i], "-n") == 0) 
        {
            if (args->number != 1){Error();}
            if ((i+1) < argc)
            {
                if (atoi(argv[i+1]) < 1)
                {
                    printf("Invalid n number, number has to be > than 0\n");
                    exit(1);
                }
                if (argv[i+1][0] != '-')
                {
                    args->number = atoi(argv[i+1]);
                    i++;
                }
            }
        }
        // If the -p argument is passed, it checks that it hasn't already been set and that a valid port number has been provided.
        else if (strcmp(argv[i], "-p") == 0) 
        {
            if (args->port != -1){Error();}
            if ((i+1) < argc)
            {
                if (atoi(argv[i+1]) < 0)
                {
                    printf("Invalid port number, number has to be >= than 0\n");
                    exit(1);
                }
                if (argv[i+1][0] != '-')
                {
                    args->port = atoi(argv[i+1]);
                    i++;
                }
            }
        }
        // If an unknown argument is passed, it prints an error message and returns 1.
        else
        {
            printf("Unknown program argument\n");
            return 1;
        }
        i++;
    }
    // If no interface is specified, it prints the available interfaces and exits.
    if (strlen(args->interface) == 0){print_interfaces();exit(0);}
    // It returns 0 if all arguments have been successfully parsed.
    return 0;
}

/**
* @brief capture packets from a network interface, based on user-specified protocols and filters
* @param args pointer to ProgramArgs struct containing user-specified command line arguments
* @return pointer to a pcap_t struct, which can be used to capture packets from the specified network interface
*/
pcap_t* work_with_device(struct ProgramArgs *args)
{
    // declare variables
    char errbuf[PCAP_ERRBUF_SIZE];
    bpf_u_int32  device_ip, netmask;
    struct bpf_program fp;
    char filter_exp[100] = {'\0'};

    // get the device IP and netmask
    if (pcap_lookupnet(args->interface, &device_ip, &netmask, errbuf) == -1)
    {
        printf("Error with  ip and netmask of device: %s\n", errbuf);
        return NULL;
    }
    // open the device for capturing
    devopen = pcap_open_live(args->interface, BUFSIZ, 1, 1000, errbuf);
    if (devopen == NULL)
    {
        printf("Error with  openning device: %s\n", errbuf);
        return NULL;
    }
    // check if the device provides Ethernet headers
    if(pcap_datalink(devopen) != DLT_EN10MB)
    {
        printf("Interface does not provide Ethernet headers");
        return NULL;
    }
    // set defaults for which protocols to capture
    if (!(args->arp || args->icmp4 || args->icmp6 || args->igmp || args->mld || args->ndp || args->tcp || args->udp))
    {
        args->arp = true;
        args->icmp4 = true;
        args->icmp6 = true;
        args->igmp = true;
        args->mld = true;
        args->ndp = true;
        args->tcp = true;
        args->udp = true;
    }
     // set the filter expression based on which protocols to capture

     // If both TCP and UDP protocols are selected and a port is specified,
     // set the filter to capture packets for the specified port on both protocols.
    if (args->tcp && args->udp) 
    {
        // If a port is not specified, set the filter to capture all ports on both protocols.
        if (args->port == -1) {sprintf(filter_exp, "(tcp and portrange 0-65535) or (udp and portrange 0-65535)");} 
        else {sprintf(filter_exp, "(tcp and port %d) or (udp and port %d)", args->port, args->port); }
    } 
     // If only TCP protocol is selected and a port is specified,
     // set the filter to capture packets for the specified port on TCP.
    else if (args->tcp) 
    {
        if (args->port == -1) {sprintf(filter_exp, "(tcp and portrange 0-65535)");} 
        else {sprintf(filter_exp, "(tcp and port %d)", args->port);}
    } 
    else if (args->udp) 
    {
        if (args->port == -1) {sprintf(filter_exp, "(udp and portrange 0-65535)");}
        else {sprintf(filter_exp, "(udp and port %d)", args->port);}
    }
    // If ARP protocol is selected, add ARP to the filter expression.
    if (args->arp) 
    {
        if (filter_exp[0] == '\0') {sprintf(filter_exp, "arp");}
        else {strcat(filter_exp, " or arp");}
    }
    // If IGMP protocol is selected, add IGMP to the filter expression.
    if (args->igmp) 
    {
        if (filter_exp[0] == '\0') {sprintf(filter_exp, "igmp");}
        else {strcat(filter_exp, " or igmp");}
    }
    // If MLD protocol is selected, add ICMPv6 MLD to the filter expression.
    // if (args->mld) {
    //     if (filter_exp[0] == '\0') { sprintf(filter_exp, "(icmp6 and (icmp6[0] >= 130 and icmp6[0] <= 132))"); }
    //     else { strcat(filter_exp, " or (icmp6 and (icmp6[0] >= 130 and icmp6[0] <= 132))"); }
    // }
    // if (args->ndp) {
    //     if (filter_exp[0] == '\0') { sprintf(filter_exp, "(icmp6 and (icmp6[0] >= 133 and icmp6[0] <= 137))"); }
    //     else { strcat(filter_exp, " or (icmp6 and (icmp6[0] >= 133 and icmp6[0] <= 137))"); }
    // }
    // If ICMPv4 protocol is selected, add ICMPv4 to the filter expression.
    if (args->icmp4) 
    {
        if (filter_exp[0] == '\0') {sprintf(filter_exp, "icmp");} 
        else {strcat(filter_exp, " or icmp");}
    }
    // Set the filter expression to capture ICMPv6 packets if args->icmp6 is true
    if (args->icmp6) 
    {
        if (filter_exp[0] == '\0') {sprintf(filter_exp, "icmp6");} 
        // Otherwise, append "or icmp6" to the existing filter expression
        else {strcat(filter_exp, " or icmp6");}
    }
    // Compile the filter expression into a filter program and apply it to the device
    if(pcap_compile(devopen, &fp, filter_exp, 0, device_ip) == -1) 
    {
        // Print an error message and return NULL if compilation fails
        printf("Error: compile failed (%s)\n", pcap_geterr(devopen));
        return NULL;
    }
    // Set the compiled filter program on the device
    if(pcap_setfilter(devopen, &fp) == -1)
	{
        // Print an error message and return NULL if setting the filter fails
		printf("Error: setting filter failed (%s)\n", pcap_geterr(devopen));
		return NULL;
	}
    // Return the device handle
    return devopen;
}

int main(int argc, char* argv[])
{
    if (argc == 1){Error();}
    struct ProgramArgs args = {
        .interface = NULL,
        .port = -1,
        .number = 1,
        .tcp = false,
        .udp = false,
        .arp = false,
        .icmp4 = false,
        .icmp6 = false,
        .igmp = false,
        .mld = false,
        .ndp = false
    };
    args.interface = malloc(20 * sizeof(char));
    int parse_result = parsing_args(argc, argv, &args);
    if (parse_result != 0)
    {
        free(args.interface);
        return 1;
    }
    else
    {
    devopen = work_with_device(&args);
    if (devopen != NULL) {
        struct pcap_pkthdr header;
        const u_char* packet;

        int k = 0;
        while (k < args.number) {
            // Get next packet from device
            packet = pcap_next(devopen, &header);
            struct ether_header* ethernet;
            ethernet = (struct ether_header*)(packet);
            char src_ip[100] = {'\0'};
            char dst_ip[100] = {'\0'};
            char src_ip6[100] = {'\0'};
            char dst_ip6[100] = {'\0'};
            char time[100] = {'\0'};
            // Format timestamp string
	        strftime(time,sizeof(time),"%H:%M:%S", localtime(&header.ts.tv_sec));
            printf("timestamp: ");
            printf("%s.%ld\n",time, (long)header.ts.tv_usec );
            uint8_t* ptr;

            // Print source MAC address
            ptr = ethernet->ether_shost;
            printf("src MAC:");
            for (int i = ETHER_ADDR_LEN; i > 0; i--) {
                if (i == ETHER_ADDR_LEN) {
                    printf(" ");
                }
                else {
                    printf(":");
                }
                printf("%02X", *ptr++);
            }
            printf("\n");


            // Print destination MAC address
            ptr = ethernet->ether_dhost;
            printf("dst MAC:");
            for (int i = ETHER_ADDR_LEN; i > 0; i--) {
                if (i == ETHER_ADDR_LEN) {
                    printf(" ");
                }
                else {
                    printf(":");
                }
                printf("%02X", *ptr++);
            }
            printf("\n");

            // Print frame length
            printf("frame length: %d bytes\n", header.len);
            // Check if Ethernet packet contains an IP packet.
            if (ntohs(ethernet->ether_type) == ETHERTYPE_IP) {
                // Extract the IP header from the packet.
                struct ip* ipHeader;
                ipHeader = (struct ip*)(packet + 14);
                // Print source and destination IP addresses
                inet_ntop(AF_INET, &ipHeader->ip_src, src_ip, INET_ADDRSTRLEN);
                inet_ntop(AF_INET, &ipHeader->ip_dst, dst_ip, INET_ADDRSTRLEN);
                printf("src IP: %s\n", src_ip);
                printf("dst IP: %s\n", dst_ip);
                // Check if the IP packet is a TCP packet.
                if (ipHeader->ip_p == IPPROTO_TCP) {
                    // Extract the TCP header from the packet.
                    struct tcphdr* tcp_header = (struct tcphdr*)(packet + 14);
                    printf("src port: %d\n", ntohs(tcp_header->th_sport));
                    printf("dst port: %d\n", ntohs(tcp_header->th_dport));
                }
                // Check if the IP packet is a UDP packet.
                else if (ipHeader->ip_p == IPPROTO_UDP) {
                    // Extract the UDP header from the packet.
                    struct udphdr* udp_header = (struct udphdr*)(packet + 14);
                    printf("src port: %d\n", ntohs(udp_header->uh_sport));
                    printf("dst port: %d\n", ntohs(udp_header->uh_dport));
                }
                // else if (ipHeader->ip_p == 1) {
                //     struct icmphdr* icmp_header = (struct icmphdr*)(packet + 14);
                // }
            }
            // Check if Ethernet packet contains an ARP packet.
            else if (ntohs(ethernet->ether_type) == ETHERTYPE_ARP) {
                // Extract the ARP header from the packet.
                struct ether_arp* arp_header = (struct ether_arp*)(packet + 14);
                // Print source and destination IP addresses.
                char arpbuf1[100] = { '\0' };
                char arpbuf2[100] = { '\0' };
                printf("src IP: %s\n", inet_ntop(AF_INET, arp_header->arp_spa, arpbuf1, 100));
                printf("dst IP: %s\n", inet_ntop(AF_INET, arp_header->arp_tpa, arpbuf2, 100));
            }
            // Check if Ethernet packet contains an IPv6 packet.
            else if (ntohs(ethernet->ether_type) == ETHERTYPE_IPV6) {
                // Extract the IPv6 header from the packet.
                struct ip6_hdr* ip6_header = (struct ip6_hdr*)(packet + 14);
                // Print source and destination IP addresses.
                inet_ntop(AF_INET6, &ip6_header->ip6_src, src_ip6, INET6_ADDRSTRLEN);
                inet_ntop(AF_INET6, &ip6_header->ip6_dst, dst_ip6, INET6_ADDRSTRLEN);
                printf("src IP: %s\n", src_ip6);
                printf("dst IP: %s\n", dst_ip6);
                // Check the protocol type of the IPv6 header, if it's TCP or UDP, extract the corresponding header information and print the source and destination ports.
                if (ip6_header->ip6_ctlun.ip6_un1.ip6_un1_nxt == IPPROTO_TCP) {
                    struct tcphdr* tcp_header = (struct tcphdr*)(packet + 14 + 40);
                    printf("src port: %d\n", ntohs(tcp_header->th_sport));
                    printf("dst port: %d\n", ntohs(tcp_header->th_dport));
                }
                else if (ip6_header->ip6_ctlun.ip6_un1.ip6_un1_nxt == IPPROTO_UDP) {
                    struct udphdr* udp_header = (struct udphdr*)(packet + 14 + 40);
                    printf("src port: %d\n", ntohs(udp_header->uh_sport));
                    printf("dst port: %d\n", ntohs(udp_header->uh_dport));
                }
                // else if (ip6_header->ip6_ctlun.ip6_un1.ip6_un1_nxt == 58) {
                //     struct icmphdr* icmp_header = (struct icmphdr*)(packet + 14);
                // }
            }
            // Print the hexadecimal and ASCII representation of the packet.
            char asciiprint[17] = { '\0' };
            int j = 0;
            while (j < (int)header.caplen) {
                // Print a new line every 16 bytes.
                if (j % 16 == 0) {
                    printf(" %s\n", asciiprint);
                    memset(asciiprint, '\0', 17);
                    printf("0x%04x ", j);
                }
                // If the byte is not printable ASCII, replace it with a dot.
                if (packet[j] < 33 || packet[j] > 127) {
                    asciiprint[j % 16] = '.';
                }
                else {
                    asciiprint[j % 16] = packet[j];
                }
                // Print the hexadecimal value of the current byte.
                printf("%02x ", packet[j]);
                j++;
            }
            // Print the last line of ASCII representation, padding with spaces as necessary.
            printf(" %*s\n\n", (16 - j % 16) * 3 + ((j % 16) ? 1 : 0), asciiprint);

            k++;
        }
    }
    // Close the device and return 0 to indicate successful completion.
    pcap_close(devopen);
    return 0;
    }
}
